homeassistant/components/homekit_controller/light.py

```diff
@@ -79,23 +80,43 @@ def color_temp(self) -> int:
         return self.service.value(CharacteristicsTypes.COLOR_TEMPERATURE)
 
     @property
-    def supported_features(self) -> int:
-        """Flag supported features."""
-        features = 0
+    def color_mode(self) -> str:
+        """Return the color mode of the light."""
+        if self.service.has(CharacteristicsTypes.HUE):
+            return COLOR_MODE_HS
+
+        if self.service.has(CharacteristicsTypes.SATURATION):
+            return COLOR_MODE_HS
+
+        if self.service.has(CharacteristicsTypes.COLOR_TEMPERATURE):
+            return COLOR_MODE_COLOR_TEMP
```

Comment:
This does not change the behavior compared to without this PR, however for lights which support hs + white modes it would be better to report according to the light's state. I did not immediately find anything useful in the library though. Any idea, @Jc2k , @bdraco?

Reply:
HomeKit (at least as defined in the public spec (https://developer.apple.com/file/?file=homekitspecification&agree=true)) doesn't have the concept of an active color mode. You can use h/s and you can use color temp. It has the characteristics you can see here, and thats it. There's no way to indicate which color mode is currently in use.

Sometimes you can figure out a bit more from the open source example code. There are some notes here:

https://github.com/apple/HomeKitADK/blob/b171c17dba6e593e243b5fdd61fb986e9fd3b5cb/HAP/HAPServiceTypes.h#L76

Again, theres no characteristic to indicate a color mode there.

https://github.com/home-assistant/core/pull/42807 (via @bdraco) says:

> The HomeKit R2 spec explicitly states that the Color Temperature char must not be used for lamps which support color.

Does that help?

Reply:
OK, I see. I'll add a comment in the code stating that's the case.

Reply:
The spec says that you're not supposed to have color and color temperature at the same time, but in practice even certified devices implement that now. Apple seems to be certifying them anyway, and the hope is that the spec will get updated in the next version to remove that restriction. iOS seems to handle it now so it seems like the documentation is behind the actual implementation

Reply:
The way iOS seems to handle it is, whichever characteristic was the last to get a notify for is the color mode you're in

Reply:
For the devices that support both color and color temp you can expect to get a notify for the color which is usually the color temp to color conversion value, and then a notify for the actual color temperature characteristic.

Some implementations will not send the notify for the color temperature to color conversion first, and only send the color temperature characteristic notify.

In practice that's likely not a problem because in the UI we already compute the approximate color of the color temperature.  It's also consistent with what I posted above which ever notify event that you get last is the color mode you are in

Reply:
The only real caveat is because events are sometimes coalesced if you receive a color and color temperature characteristic in the same event, you should prefer the color temp characteristic for the color mode

Reply:
That'll be awkward to implement with the current abstraction.

For a start, aiohomekit doesn't do any tracking of when the last change occurred.

Even if it did, polling and notifications get funnelled into the same state machine and right now we don't know which is which. And there is always polling as it was largely consequence free and a good way to recover on networks that have flakey wifi.

If conversions are lossy we would potentially see a notification for one colour mode. Then after a minute there would be a poll of all colour values, which might look like a change into a different colour mode.

There's also the gotcha that we can't recover the correct colour mode after a restart. We'd have to wait for an event.

We'd also need to feed in our own colour mode changes that we generate, as you don't get a notify for a change you made through your own connection.

At the moment this sounds like a workaround and I'm not keen on going there unless it's a massive problem to not go there.

Reply:
There are only a few that implement both so I think it's fine to not support that setup. It does seem to be getting more common though so probably something for the future

Reply:
Hopefully we'll get a new spec or an ADK update... (doubt it).

It might be less sucky after I've done all the refactoring I want to in aiohomekit (I want the abstraction to be higher level than it is right now).

Reply:
Based on your discussion I don't think the simplification is correct, it should instead be like this:
- If the light reports support for both ct and hs, we should expose that as `supported_color_modes = {COLOR_MODE_COLOR_TEMP, COLOR_MODE_HS}` (no change compared to before this PR)
- Because aiohomekit does not track which color mode the light is in, a light supporting both ct and hs should just report color_mode `COLOR_MODE_HS` (no change compared to this PR)

Reply:
To be more pedantic on the second bullet: the HAP protocol itself doesn't really have a concept of a current colour mode to track and aiohomekit's abstraction doesn't let us work around that. And even if it did, it would be an incomplete work around (we would lose the "current colour mode" on a simple HA restart for example). I assume the official iOS implementation is subject to the same limitations too.

(The thing I'm being pedantic about is that despite the changes in iOS 13 it's still not the case there is a variable /to/ track, we would have to track multiple variables for multiple event types (push, put and poll) and then infer our own value from the event timestamps of a subset of event types, only to have a partially working implementation at the end of it).

In the future we could implement the heuristics described to make it closer to the ideal, but short of Apple adding a new colour mode characteristic there is a potential for our implementation to seem "glitchy". So we might not want to implement such heuristics at all.

But the ultimate conclusion is the same, I think.

Reply:
OK, updated again.

Reply:
@bdraco, @Jc2k Is this PR OK now, or are more changes needed?